File: api/app/routes/matching.py
from __future__ import annotations

import asyncio
import time

from fastapi import APIRouter, Depends, HTTPException, Request, status
from slowapi import Limiter
from slowapi.util import get_remote_address
from sqlalchemy.orm import Session

from app import flags, metrics, schemas
from app.database import get_db
from app.logger import get_logger
from app.models import CrewProfile, Job, JobHistoryEntry, User
from app.security import require_admin_session
from app.services.matching_engine import (
    CandidateProfile,
    JobSummary,
    match_candidate_to_jobs,
)
from app.settings import settings

_limiter = Limiter(key_func=get_remote_address)
log = get_logger("carver.matching")

router = APIRouter(prefix="/matching", tags=["matching"], dependencies=[Depends(require_admin_session)])


def _user_to_candidate(user: User, profile: CrewProfile | None = None, job_history: list[JobHistoryEntry] | None = None) -> CandidateProfile:
    """Build a CandidateProfile from a User row, optionally enriched with CrewProfile data."""
    certs: list[str] = []
    langs: list[str] = []
    history: list[dict[str, str]] = []
    bio = ""

    if profile:
        if profile.certifications:
            certs = [c.strip() for c in profile.certifications.replace("\n", ",").split(",") if c.strip()]
        if profile.languages:
            langs = [lang.strip() for lang in profile.languages.split(",") if lang.strip()]
        bio = profile.bio or ""

    if job_history:
        for e in job_history:
            history.append({
                "role": e.role,
                "yacht": e.yacht_name,
                "yacht_type": e.yacht_type or "",
                "start_date": e.start_date or "",
                "end_date": e.end_date or "",
                "description": (e.description or "")[:200],
            })

    return CandidateProfile(
        user_key=str(user.id),
        first_name=profile.first_name or "" if profile else user.full_name.split()[0] if user.full_name else "",
        last_name=profile.last_name or "" if profile else " ".join(user.full_name.split()[1:]) if user.full_name else "",
        sex=profile.sex or "" if profile else "",
        desired_role=profile.desired_role or user.role or "" if profile else user.role or "",
        location=profile.current_location or user.current_location or "" if profile else user.current_location or "",
        preferred_locations=profile.preferred_locations or "" if profile else "",
        nationality=profile.nationality or user.nationality or "" if profile else user.nationality or "",
        years_experience=profile.years_experience or str(user.years_experience or "") if profile else str(user.years_experience or ""),
        salary_min=profile.salary_min or "" if profile else "",
        salary_max=profile.salary_max or "" if profile else "",
        contract_type=profile.contract_type or "" if profile else "",
        rotation_preference=profile.rotation_preference or "" if profile else "",
        available_from=profile.available_from or "" if profile else "",
        certifications=certs,
        languages=langs,
        bio=bio,
        job_history=history,
    )


def _job_to_summary(j: Job) -> JobSummary:
    return JobSummary(
        job_id=j.id,
        title=j.title,
        role=j.role or "",
        department=j.department or "",
        location=j.location,
        yacht_type=j.yacht_type or "",
        yacht_length_m=j.yacht_length_m,
        start_date=j.start_date or "",
        contract_type=j.contract_type or "",
        rotation=j.rotation or "",
        season=j.season or "",
        salary_min=j.salary_min,
        salary_max=j.salary_max,
        salary_currency=j.salary_currency or "EUR",
        experience_required_years=j.experience_required_years,
        certifications_required=j.certifications_required or "",
        languages_required=j.languages_required or "",
        description=j.description or "",
    )


@router.post("/run", response_model=schemas.MatchingRunResponse)
@_limiter.limit("10/minute")
async def run_match(request: Request, payload: schemas.MatchingRequest, db: Session = Depends(get_db)):
    if not flags.is_enabled("matching"):
        metrics.increment("feature_blocked")
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Job matching is temporarily disabled.")

    if not settings.OPENAI_API_KEY:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Matching not configured (missing API key).")

    log.info("Run request | user_id=%d | job_statuses=%s", payload.user_id, payload.job_statuses)

    user = db.query(User).filter(User.id == payload.user_id, User.is_active.is_(True)).first()
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

    all_jobs = (
        db.query(Job)
        .filter(Job.status.in_(payload.job_statuses))
        .order_by(Job.created_at.desc())
        .all()
    )
    if not all_jobs:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="No jobs found")

    profile = db.query(CrewProfile).filter(CrewProfile.user_key == str(user.id)).first()
    job_history = None
    if profile:
        job_history = (
            db.query(JobHistoryEntry)
            .filter(JobHistoryEntry.user_key == str(user.id))
            .order_by(JobHistoryEntry.start_date.desc())
            .limit(10)
            .all()
        )

    candidate = _user_to_candidate(user, profile, job_history)
    job_summaries = [_job_to_summary(j) for j in all_jobs]

    log.info("Running match | user_id=%d | jobs=%d", payload.user_id, len(all_jobs))
    t0 = time.perf_counter()

    try:
        results = await asyncio.to_thread(
            match_candidate_to_jobs,
            api_key=settings.OPENAI_API_KEY,
            model=settings.OPENAI_MODEL,
            candidate=candidate,
            jobs=job_summaries,
        )
    except Exception as exc:
        log.error("Match engine error | user_id=%d | %s", payload.user_id, exc)
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Matching failed. Please try again.")
    finally:
        metrics.record_ai_response_time(round((time.perf_counter() - t0) * 1000))

    metrics.increment("matching_completed")
    log.info("Match complete | user_id=%d | results=%d | matched=%d",
             payload.user_id, len(results), sum(1 for r in results if r.matched))

    return schemas.MatchingRunResponse(
        request_id=f"match-{payload.user_id}-{int(time.time())}",
        matches=[
            schemas.MatchResultItem(
                job_id=str(r.job_id),
                matched=r.matched,
                compatibility=r.compatibility,
                reason=r.reason,
                strengths=r.strengths,
                gaps=r.gaps,
                factor_scores=r.factor_scores,
            )
            for r in results
        ],
    )

File: jobcarver-go/internal/handlers/matching/matching.go
// Package matching ports api/app/routes/matching.py: the admin-gated /matching
// endpoint that scores every open/priority job against a user and spends one
// token. The actual scoring runs through internal/matching (which routes the
// model call through the internal/ai stub).
package matching

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"jobcarver/internal/config"
	"jobcarver/internal/credits"
	"jobcarver/internal/db"
	"jobcarver/internal/flags"
	"jobcarver/internal/httpx"
	matchengine "jobcarver/internal/matching"
	"jobcarver/internal/metrics"
	"jobcarver/internal/models"
	"jobcarver/internal/security"
)

// Register mounts POST /matching/run (admin session required, mirroring the
// require_admin_session dependency on the Python router).
//
// NOTE: the /matching prefix is shared with the crewmatch handler, so this uses
// r.Group + a full path rather than r.Route("/matching") to avoid a chi mount
// collision.
func Register(r chi.Router) {
	r.Group(func(r chi.Router) {
		r.Use(security.RequireSession)
		r.Use(adminOnly)
		r.Post("/matching/run", run)
	})
}

func adminOnly(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		sess := security.SessionFromContext(r.Context())
		if sess == nil || sess.Role != "admin" {
			httpx.WriteError(w, http.StatusForbidden, "Admin access required", httpx.CRV2005)
			return
		}
		next.ServeHTTP(w, r)
	})
}

type runRequest struct {
	UserID         int      `json:"user_id"`
	JobLimit       int      `json:"job_limit"`
	JobStatuses    []string `json:"job_statuses"`
	TimeoutSeconds int      `json:"timeout_seconds"`
}

type matchResultItem struct {
	JobID         string             `json:"job_id"`
	Matched       bool               `json:"matched"`
	Compatibility float64            `json:"compatibility"`
	Reason        string             `json:"reason"`
	Strengths     []string           `json:"strengths"`
	Gaps          []string           `json:"gaps"`
	FactorScores  map[string]float64 `json:"factor_scores"`
}

type runResponse struct {
	RequestID        string            `json:"request_id"`
	CreditsRemaining int               `json:"credits_remaining"`
	Matches          []matchResultItem `json:"matches"`
}

func run(w http.ResponseWriter, r *http.Request) {
	if !flags.IsEnabled("matching") {
		metrics.Increment("feature_blocked")
		httpx.WriteError(w, http.StatusServiceUnavailable,
			"Job matching is temporarily disabled.", httpx.CRV1008)
		return
	}

	if strings.TrimSpace(config.S.OpenAIAPIKey) == "" {
		httpx.WriteError(w, http.StatusServiceUnavailable,
			"Matching not configured (missing API key).", httpx.CRV3001)
		return
	}

	var body runRequest
	if err := httpx.Decode(r, &body); err != nil || body.UserID < 1 {
		httpx.WriteError(w, http.StatusUnprocessableEntity, "Invalid request body.", httpx.CRV1003)
		return
	}
	statuses := body.JobStatuses
	if len(statuses) == 0 {
		statuses = []string{"open", "priority"}
	}

	var user models.User
	if err := db.DB.Where("id = ? AND is_active = ?", body.UserID, true).First(&user).Error; err != nil {
		httpx.WriteError(w, http.StatusNotFound, "User not found", httpx.CRV1005)
		return
	}

	var jobs []models.Job
	if err := db.DB.Where("status IN ?", statuses).Order("created_at DESC").Find(&jobs).Error; err != nil {
		httpx.WriteError(w, http.StatusInternalServerError, "Query failed", httpx.CRV1002)
		return
	}
	if len(jobs) == 0 {
		httpx.WriteError(w, http.StatusNotFound, "No jobs found", httpx.CRV1005)
		return
	}

	// Spend one token up front; refund on engine error.
	if err := credits.Spend(db.DB, user.Email, 1); err != nil {
		httpx.WriteError(w, http.StatusPaymentRequired,
			"User does not have enough credits to run matching.", httpx.CRV1004)
		return
	}
	creditsRemaining := credits.GetBalance(db.DB, user.Email)

	userKey := fmt.Sprintf("%d", user.ID)

	var profile models.CrewProfile
	hasProfile := db.DB.Where("user_key = ?", userKey).First(&profile).Error == nil

	var history []models.JobHistoryEntry
	if hasProfile {
		db.DB.Where("user_key = ?", userKey).Order("start_date DESC").Limit(10).Find(&history)
	}

	var docs []models.Document
	db.DB.Where("user_key = ? AND scanned_text IS NOT NULL", userKey).Find(&docs)
	var docParts []string
	for _, d := range docs {
		if d.ScannedText != nil && *d.ScannedText != "" {
			docParts = append(docParts, "["+strings.ToUpper(d.DocType)+"] "+*d.ScannedText)
		}
	}
	docSummary := strings.Join(docParts, "\n\n")

	candidate := userToCandidate(user, hasProfile, profile, history, docSummary)
	summaries := make([]matchengine.JobSummary, 0, len(jobs))
	for _, j := range jobs {
		summaries = append(summaries, jobToSummary(j))
	}

	t0 := time.Now()
	results := matchengine.MatchCandidateToJobs(r.Context(), config.S.OpenAIModel, candidate, summaries, 0, nil)
	metrics.RecordAIResponseTime(int(time.Since(t0).Milliseconds()))

	metrics.Increment("matching_completed")

	items := make([]matchResultItem, 0, len(results))
	for _, res := range results {
		strengths := res.Strengths
		if strengths == nil {
			strengths = []string{}
		}
		gaps := res.Gaps
		if gaps == nil {
			gaps = []string{}
		}
		fs := res.FactorScores
		if fs == nil {
			fs = map[string]float64{}
		}
		items = append(items, matchResultItem{
			JobID:         fmt.Sprintf("%d", res.JobID),
			Matched:       res.Matched,
			Compatibility: res.Compatibility,
			Reason:        res.Reason,
			Strengths:     strengths,
			Gaps:          gaps,
			FactorScores:  fs,
		})
	}

	httpx.JSON(w, http.StatusOK, runResponse{
		RequestID:        fmt.Sprintf("match-%d-%d", body.UserID, time.Now().Unix()),
		CreditsRemaining: creditsRemaining,
		Matches:          items,
	})
}

// userToCandidate ports _user_to_candidate.
func userToCandidate(user models.User, hasProfile bool, profile models.CrewProfile, history []models.JobHistoryEntry, docSummary string) matchengine.CandidateProfile {
	var certs, langs []string
	var bio string
	if hasProfile {
		if profile.Certifications != nil {
			for _, c := range strings.Split(strings.ReplaceAll(*profile.Certifications, "\n", ","), ",") {
				if c = strings.TrimSpace(c); c != "" {
					certs = append(certs, c)
				}
			}
		}
		if profile.Languages != nil {
			for _, l := range strings.Split(*profile.Languages, ",") {
				if l = strings.TrimSpace(l); l != "" {
					langs = append(langs, l)
				}
			}
		}
		bio = deref(profile.Bio)
	}

	hist := make([]map[string]string, 0, len(history))
	for _, e := range history {
		hist = append(hist, map[string]string{
			"role":        e.Role,
			"yacht":       e.YachtName,
			"yacht_type":  deref(e.YachtType),
			"start_date":  deref(e.StartDate),
			"end_date":    deref(e.EndDate),
			"description": clip(deref(e.Description), 200),
		})
	}

	firstName, lastName := splitName(user.FullName)
	desiredRole := user.Role
	location := deref(user.CurrentLocation)
	nationality := deref(user.Nationality)
	yearsExp := ""
	if user.YearsExperience != nil {
		yearsExp = fmt.Sprintf("%d", *user.YearsExperience)
	}

	if hasProfile {
		if v := deref(profile.FirstName); v != "" {
			firstName = v
		}
		if v := deref(profile.LastName); v != "" {
			lastName = v
		}
		if v := deref(profile.DesiredRole); v != "" {
			desiredRole = v
		}
		if v := deref(profile.CurrentLocation); v != "" {
			location = v
		}
		if v := deref(profile.Nationality); v != "" {
			nationality = v
		}
		if v := deref(profile.YearsExperience); v != "" {
			yearsExp = v
		}
	}

	c := matchengine.CandidateProfile{
		UserKey:         fmt.Sprintf("%d", user.ID),
		FirstName:       firstName,
		LastName:        lastName,
		DesiredRole:     desiredRole,
		Location:        location,
		Nationality:     nationality,
		YearsExperience: yearsExp,
		Certifications:  certs,
		Languages:       langs,
		Bio:             bio,
		JobHistory:      hist,
		DocumentSummary: docSummary,
	}
	if hasProfile {
		c.Sex = deref(profile.Sex)
		c.PreferredLocations = deref(profile.PreferredLocations)
		c.SalaryMin = deref(profile.SalaryMin)
		c.SalaryMax = deref(profile.SalaryMax)
		c.ContractType = deref(profile.ContractType)
		c.RotationPreference = deref(profile.RotationPreference)
		c.AvailableFrom = deref(profile.AvailableFrom)
	}
	return c
}

// jobToSummary ports _job_to_summary.
func jobToSummary(j models.Job) matchengine.JobSummary {
	currency := "EUR"
	if j.SalaryCurrency != nil && *j.SalaryCurrency != "" {
		currency = *j.SalaryCurrency
	}
	status := j.Status
	if status == "" {
		status = "open"
	}
	return matchengine.JobSummary{
		JobID:                   j.ID,
		Title:                   j.Title,
		Role:                    j.Role,
		Department:              deref(j.Department),
		Location:                j.Location,
		YachtType:               deref(j.YachtType),
		YachtLengthM:            j.YachtLengthM,
		StartDate:               deref(j.StartDate),
		ContractType:            deref(j.ContractType),
		Rotation:                deref(j.Rotation),
		Season:                  deref(j.Season),
		SalaryMin:               j.SalaryMin,
		SalaryMax:               j.SalaryMax,
		SalaryCurrency:          currency,
		ExperienceRequiredYears: j.ExperienceRequiredYears,
		CertificationsRequired:  deref(j.CertificationsRequired),
		LanguagesRequired:       deref(j.LanguagesRequired),
		Description:             deref(j.Description),
		Status:                  status,
	}
}

func deref(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}

func clip(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

func splitName(full string) (string, string) {
	full = strings.TrimSpace(full)
	if full == "" {
		return "", ""
	}
	parts := strings.Fields(full)
	if len(parts) == 1 {
		return parts[0], ""
	}
	return parts[0], strings.Join(parts[1:], " ")
}
